[PD][Feat]: add SJF (Shortest Job First) scheduling policy

qianlihuang opened the pull request:
## Motivation

In disaggregated serving (PD disaggregation) architecture, the prefill node acts as a producer that processes input tokens before handing off requests to decode nodes (consumers). When multiple requests are queued on a prefill-only node, the scheduling order significantly impacts the **Time-To-First-Token (TTFT)** distribution and the overall flow efficiency.

Currently, SGLang supports policies like `fcfs`, `lpm`, `dfs-weight`, `lof`, `random`, and `priority`. However, none of these specifically optimize for **minimizing average/mean TTFT** or maximizing request turnover rate.

**Why SJF (Shortest Job First)?**

In real-world production scenarios, particularly **multi-turn conversations**, input token lengths typically exhibit a **long-tail (geometric-like) distribution**. There is usually a high volume of short, early-turn requests compared to a sparse number of long-context, later-turn requests.

Under the current FCFS policy, this distribution leads to significant **Head-of-Line (HOL) blocking**: a few long-context requests can block the execution of numerous subsequent short requests. This not only inflates waiting times but also creates a "stop-and-go" flow to the decode nodes.

By implementing **SJF**, which prioritizes requests with shorter input lengths, we can:
- **Eliminate the blocking effect** caused by sparse long-context requests on dense short requests.
- **Minimize the average waiting time** and achieve significantly better Mean TTFT metrics.
- **Boost the Request Throughput (RPS)** of the prefill node, enabling a higher volume of requests to be produced and handed off to decode nodes for consumption per unit time.
- Improve the overall responsiveness and user experience for the majority of queries.

<img width="726" height="516" alt="image" src="https://github.com/user-attachments/assets/7b438c8b-cbce-49a1-bb46-d68a1fb9843e" />


<img width="1235" height="436" alt="image" src="https://github.com/user-attachments/assets/5760d0ec-3bd0-4e8f-ba0c-77613c4d450c" />

related: https://sgl-fru7574.slack.com/archives/C08AP4WU8P3/p1765348436725869?thread_ts=1764929650.264799&cid=C08AP4WU8P3


## Modifications

### Core Changes
1. **`python/sglang/srt/managers/schedule_policy.py`**
   - Added `SJF = "sjf"` to `CacheAgnosticPolicy` enum
   - Added SJF branch in `calc_priority()` method
   - Implemented `_sort_by_shortest_job()` static method that sorts requests by `len(origin_input_ids)`
   - Supports combination with priority scheduling (priority first, then SJF within same priority)

2. **`python/sglang/srt/server_args.py`**
   - Added `"sjf"` to `--schedule-policy` choices

### Tests
3. **`test/srt/test_sjf_scheduling.py`**
   - Unit tests for basic SJF scheduling
   - Unit tests for SJF combined with priority scheduling

## Usage

```bash
# For prefill-only node in PD disaggregation
python3 -m sglang.launch_server \
  --model <model_path> \
  --schedule-policy sjf \
  --disaggregation-mode prefill \
  ...
```
  
## Checklist
- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

## Related Issues
N/A (New feature proposal)

abcdea committed:
feat(scheduler): add SJF (Shortest Job First) scheduling policy

Add SJF scheduling policy for prefill-only nodes in disaggregated
serving (PD separation). This policy prioritizes shorter input requests
to minimize average TTFT by reducing overall queue waiting time.

- Add SJF enum value to CacheAgnosticPolicy
- Implement _sort_by_shortest_job method
- Add sjf option to --schedule-policy argument
- Add unit tests for SJF scheduling

abcdea committed:
feat: add SJF scheduling policy for prefill nodes

- Add SJF (Shortest Job First) scheduling policy to minimize average TTFT in PD separation
- Implement _sort_by_shortest_job method that sorts requests by input length
- Support SJF with priority scheduling combination

Motivation: In PD separation mode, prefill nodes can achieve better mean TTFT by prioritizing short requests, reducing overall queue waiting time.

abcdea committed:
docs: update comments to use 'disaggregation' instead of 'separation'

- Update comments in schedule_policy.py and test_sjf_scheduling.py
- Change 'PD separation' to 'PD disaggregation' for consistency

abcdea committed:
docs: add 'sjf' option to --schedule-policy parameter documentation

- Update server_arguments.md to include 'sjf' in the list of available scheduling policies

ShangmingCai reviewed (commented):
please move test/srt/test_sjf_scheduling.py to test/manual/test_sjf_scheduling.py

abcdea committed:
test: move SJF scheduling test to manual directory

Move test_sjf_scheduling.py from test/srt/ to test/manual/ as per reviewer suggestion

ShangmingCai reviewed (approved):
Policy LGTM. But not sure about the test actually (since this test is a manual test, I am OK with it).

qianlihuang committed:
Merge branch 'main' into feature/sjf-scheduling-policy

qianlihuang committed:
Merge branch 'main' into feature/sjf-scheduling-policy

abcdea committed:
Update priority scheduling validation to include 'sjf' policy

abcdea committed:
Merge branch 'feature/sjf-scheduling-policy' of https://github.com/qianlihuang/sglang into feature/sjf-scheduling-policy

qianlihuang committed:
Merge branch 'main' into feature/sjf-scheduling-policy

qianlihuang committed:
Merge branch 'main' into feature/sjf-scheduling-policy
